config-system.lisp

```diff
@@ -0,0 +1,157 @@
+
+(defun %generate-config-var-code (declare-type name default validator documentation)
+  (check-type documentation string)
+  (with-gensyms (default-value is-valid func)
+    `(progn
+       (,declare-type ,name ,default ,documentation)
+       (let* ((,default-value ,name)
+              (,func ,validator)
+              (,is-valid (funcall ,func ,default-value)))
+         (if ,is-valid
+             (setf (gethash (quote ,name) *config-vars*)
+                   (make-config-info :name (quote ,name) :default ,default-value
+                                     :validator ,func :doc ,documentation))
+             (error 'invalid-datum-error :place (quote ,name) :value ,default-value))))))
+
+(defmacro define-config-parameter (name default validator documentation)
+  "Create and register a configurable variable with the given default value,
+validator function, and documentation using DEFPARAMETER"
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defparameter name default validator documentation))
+
+(defmacro define-config-var (name default validator documentation)
+  "Create and register a configurable variable with the given default value,
+validator function, and documentation using DEFVAR"
+  ;; if we are given a lambda, we need to evaluate it in order to call it
+  (%generate-config-var-code 'defvar name default validator documentation))
+
+(defmacro define-config-enum (name default values documentation)
+  "Same as DEFINE-CONFIG-PARAMETER, except the validation function
+is built by creating a function that checks if the set  value is in the VALUES list."
+  ;; if we wanted to get fancy, we could define a special type for this enumeration at this point too.
+  `(define-config-parameter ,name ,default (lambda (x) (member x ,values :test #'equal))
+                            ,documentation))
+
+(defun list-all-configurations ()
+  "List all of the available customizable settings"
+  (hash-table-values *config-vars*))
+
+(defun %find-possible-settings (setting-name table)
+  "Find the settings that have the same symbol name of SETTING-NAME but are in a different package"
+  (declare (type hash-table table)
+           (type symbol setting-name))
+  (let ((name (symbol-name setting-name)))
+    ;; for some reason, the package isn't always included with the symbol name even
+    ;; if the top level isn't in the symbol's package, so we need to manually get the symbol name.
+    (mapcar #'%full-symbol-string
+            (remove-if-not (lambda (x) (string-equal (symbol-name x) name))
+                           (mapcar #'config-info-name (hash-table-values table))))))
+
+(defun get-configuration-info (setting-name)
+  "Find the info for the config variable stored in the symbol SETTING-NAME."
+  (declare (type symbol setting-name))
+  (gethash setting-name *config-vars*))
+
+(defmacro %set-configuration (setting-name value)
```

Comment:
I think it would be a good idea to store/dispatch based not only on the symbol but on the symbols package as well. This macro and the underlying hash table don't appear to currently support this. I solved this by making the hash table use `#'equalp` and using `(package . symbol)` for the hash key. What do you think of this? alternatively one could use the full name of the symbol as the key. This work is already partially done within `%full-symbol-string`.

Reply:
Symbols from different packages are distinct items, so there is no need to do this: they will never compare to be equal. For example, while in the `cl-user` package, the form `(eql `sb-ext::%boundp '%boundp)` returns `nil`. So as long as a package isn't using a symbol from another package, symbols with the same name from different packages will be distinct keys in the hash table.

Reply:
Ah! thanks for explaining this! looks like I have solved a nonexistent problem :rofl: